probe-rs/src/debug/variable.rs

```diff
@@ -1,41 +1,362 @@
+use super::*;
+use crate::Error;
+use anyhow::anyhow;
+use gimli::{DebugInfoOffset, UnitOffset};
 use num_traits::Zero;
-use thousands::Separable;
+use std::{
+    cell::{Cell, RefCell},
+    cmp::Ordering,
+};
 
-use super::*;
-use std::{convert::TryInto, fmt};
-
-/// VariableKind is a tag used to differentiate the nature of a variable. The DAP protocol requires a differentiation between 'Named' and 'Indexed'. We've added some flags to control when variables require unique handling or decoding the value during runtimeprocessing.
-#[derive(Debug, Clone, PartialEq)]
-pub enum VariableKind {
-    /// An Indexed variable (bound to an ordinal position), such as the sequenced members of an Array or Vector
-    Indexed,
-    /// A variable that is identified by it's name, and is not bound to a specific ordinal position.
-    Named,
-    /// A variable that is the target of a pointer variable
-    Referenced,
-    /// As the default, his should never be the final value for a Variable
-    Undetermined,
+/// VariableCache stores every available `Variable`, and provides methods to create and navigate the parent-child relationships of the Variables.
+/// There should be ONLY ONE `VariableCache` per `DebugInfo`. Because of the multiple ways in which it is updated, all references to `VariableCache` are *immutable*, and can only be updated through its methods, which provide *interior mutability"
```

Comment:
No it does not make sense for it to be re-used. As per the 'ONLY ONE' comment above. At different execution points, there will be different variables in scope and their values may have changed relative to prior execution points. For this reason, every unwind action (stop at a specific execution point) will rebuild the cache with ONLY the relevant data.

Reply:
i think 'ONLY ONE' is not really clear, since that implies to me that it will be re-used. Otherwise, multiple variable caches are created, and then it's not 'ONLY ONE'.

Maybe "only one at the same time" would make it more clear?

Reply:
OK. I've updated the comment to be more exact about the relationship between VariableCache and DebugInfo